gradle/artifacts.gradle

```diff
@@ -0,0 +1,64 @@
+if (project.plugins.hasPlugin('com.android.library')) {
+    android.libraryVariants.all { variant ->
+        Task javadocTask = task("generate${variant.name.capitalize()}Javadoc", type: Javadoc) {
+            group = 'artifact'
+            description "Generates Javadoc for $variant.name"
+
+            // Source files from the variant
+            source = variant.javaCompiler.source
+
+            // Classpath from the variant + android.jar
+            String androidJar = "${android.sdkDirectory}/platforms/${android.compileSdkVersion}/android.jar"
+            classpath = variant.javaCompiler.classpath + files(androidJar)
+
+            // The Android online reference doesn't include package-list, so we have to use the local one
+            String packageListRef = "${android.sdkDirectory}/docs/reference/"
+            options.linksOffline 'http://d.android.com/reference/', packageListRef
+
+            // Additional links for any RxJava references
+            options.links 'http://reactivex.io/RxJava/javadoc/'
+
+            // Exclude generated files
+            exclude '**/BuildConfig.java'
+            exclude '**/R.java'
+
+            // Output to a unique javadoc folder per variant
+            destinationDir = new File(project.docsDir, "javadoc-$variant.name")
+        }
+
+        // For official releasese, don't prefix the name so the artifact is published correctly
+        // (Can't seem to modify it for publishing, for whatever reason...)
+        String classifierPrefix = (variant.name == 'release') ? '' : "$variant.name-"
+
+        Task javadocJarTask = task("generate${variant.name.capitalize()}JavadocJar", type: Jar, dependsOn: javadocTask) {
+            group = 'artifact'
+            description = "Generates Javadoc jar for $variant.name"
+
+            classifier = "${classifierPrefix}javadoc"
+            from javadocTask.destinationDir
+        }
+
+        Task sourcesJarTask = task("generate${variant.name.capitalize()}SourcesJar", type: Jar) {
+            group = 'artifact'
+            description = "Generates sources jar for $variant.name"
+
+            classifier = "${classifierPrefix}sources"
+            from variant.javaCompiler.source
```

Comment:
No. `variant.java.compile` is deprecated. From [BaseVariant.java](https://android.googlesource.com/platform/tools/base/+/master/build-system/gradle-core/src/main/groovy/com/android/build/gradle/api/BaseVariant.java):

```java
/**
 * Returns the Java Compilation task if javac was configured to compile the source files.
 * @deprecated prefer {@link #getJavaCompiler} which always return the java compiler task
 * irrespective of which tool chain (javac or jack) used.
 */
@Nullable
@Deprecated
JavaCompile getJavaCompile() throws IllegalStateException;

/**
 * Returns the Java Compiler task which can be either javac or jack depending on the project
 * configuration.
 */
@NonNull
AbstractCompile getJavaCompiler();
```